File: packages/effects/src/resample.ts
// Resampling for the two-tone pipeline (SPEC 5.4, 10): the padded crop, the cover fit to 800 by
// 450 and the nearest-neighbour upscale.
//
// The pinned Lanczos3 is Pillow's resampler, reproduced step for step (libImaging/Resample.c):
// the kernel sinc(x) sinc(x / 3) on |x| < 3, support scaled by the downscale factor, taps
// centered at in0 + (xx + 0.5) * scale and rounded with (int)(v + 0.5), coefficients normalized
// to one and then quantized to 22-bit fixed point, a horizontal pass and a vertical pass each
// through 8-bit storage, and the sum rounded with a half-unit bias and floored. The three Lanczos
// implementations measured in the slides experiment disagreed on about 120 cells of 1.44 million
// (slides report section 3.4); pinning Pillow's arithmetic makes the twins the rounds approved
// reproducible, and gives the Rust crate one integer specification to match.
import type { Box4, BitImage, GrayImage } from './image.ts';

const SUPPORT = 3;
const PRECISION_BITS = 32 - 8 - 2;
const PRECISION = 2 ** PRECISION_BITS;
const HALF = 2 ** (PRECISION_BITS - 1);

function sinc(x: number): number {
  if (x === 0) return 1;
  const px = x * Math.PI;
  return Math.sin(px) / px;
}

function lanczos(x: number): number {
  if (x >= -3 && x < 3) return sinc(x) * sinc(x / 3);
  return 0;
}

/**
 * The per-axis tap table: `ksize` coefficients per output pixel in `kk`, and per output pixel its
 * first input index and tap count in `bounds`.
 */
export type Coeffs = { ksize: number; bounds: Int32Array; kk: Int32Array };

/** Pillow's precompute_coeffs plus normalize_coeffs_8bpc, for one axis. */
export function precomputeCoeffs(
  inSize: number,
  in0: number,
  in1: number,
  outSize: number,
): Coeffs {
  const scale = (in1 - in0) / outSize;
  const filterscale = scale < 1 ? 1 : scale;
  const support = SUPPORT * filterscale;
  const ksize = Math.ceil(support) * 2 + 1;
  const bounds = new Int32Array(outSize * 2);
  const kk = new Int32Array(outSize * ksize);
  const k = new Float64Array(ksize);
  const ss = 1 / filterscale;
  for (let xx = 0; xx < outSize; xx += 1) {
    const center = in0 + (xx + 0.5) * scale;
    let xmin = Math.trunc(center - support + 0.5);
    if (xmin < 0) xmin = 0;
    let xmax = Math.trunc(center + support + 0.5);
    if (xmax > inSize) xmax = inSize;
    xmax -= xmin;
    let ww = 0;
    for (let x = 0; x < xmax; x += 1) {
      const w = lanczos((x + xmin - center + 0.5) * ss);
      k[x] = w;
      ww += w;
    }
    for (let x = 0; x < xmax; x += 1) if (ww !== 0) k[x] = (k[x] ?? 0) / ww;
    for (let x = xmax; x < ksize; x += 1) k[x] = 0;
    for (let x = 0; x < ksize; x += 1) {
      const v = k[x] ?? 0;
      kk[xx * ksize + x] =
        v < 0 ? Math.trunc(-0.5 + v * PRECISION) : Math.trunc(0.5 + v * PRECISION);
    }
    bounds[xx * 2] = xmin;
    bounds[xx * 2 + 1] = xmax;
  }
  return { ksize, bounds, kk };
}

function clip8(ss: number): number {
  const v = Math.floor(ss / PRECISION);
  return v < 0 ? 0 : v > 255 ? 255 : v;
}

function horizontalPass(
  src: GrayImage,
  offset: number,
  outH: number,
  outW: number,
  c: Coeffs,
): GrayImage {
  const out = new Uint8Array(outW * outH);
  const { ksize, bounds, kk } = c;
  for (let yy = 0; yy < outH; yy += 1) {
    const inRow = (yy + offset) * src.width;
    const outRow = yy * outW;
    for (let xx = 0; xx < outW; xx += 1) {
      const xmin = bounds[xx * 2] ?? 0;
      const xmax = bounds[xx * 2 + 1] ?? 0;
      const kBase = xx * ksize;
      let ss = HALF;
      for (let x = 0; x < xmax; x += 1)
        ss += (src.data[inRow + x + xmin] ?? 0) * (kk[kBase + x] ?? 0);
      out[outRow + xx] = clip8(ss);
    }
  }
  return { width: outW, height: outH, data: out };
}

function verticalPass(src: GrayImage, outH: number, c: Coeffs): GrayImage {
  const w = src.width;
  const out = new Uint8Array(w * outH);
  const { ksize, bounds, kk } = c;
  for (let yy = 0; yy < outH; yy += 1) {
    const ymin = bounds[yy * 2] ?? 0;
    const ymax = bounds[yy * 2 + 1] ?? 0;
    const kBase = yy * ksize;
    const outRow = yy * w;
    for (let xx = 0; xx < w; xx += 1) {
      let ss = HALF;
      for (let y = 0; y < ymax; y += 1)
        ss += (src.data[(y + ymin) * w + xx] ?? 0) * (kk[kBase + y] ?? 0);
      out[outRow + xx] = clip8(ss);
    }
  }
  return { width: w, height: outH, data: out };
}

/**
 * Pillow's Image.resize(size, LANCZOS, box) for an 8-bit image. The box is in source pixels and
 * may carry fractions, as ImageOps.fit passes it.
 */
export function resizeLanczos3(src: GrayImage, dstW: number, dstH: number, box?: Box4): GrayImage {
  const [x0, y0, x1, y1] = box ?? [0, 0, src.width, src.height];
  if (
    dstW === src.width &&
    dstH === src.height &&
    x0 === 0 &&
    y0 === 0 &&
    x1 === src.width &&
    y1 === src.height
  ) {
    return { width: src.width, height: src.height, data: new Uint8Array(src.data) };
  }
  const horiz = precomputeCoeffs(src.width, x0, x1, dstW);
  const vert = precomputeCoeffs(src.height, y0, y1, dstH);
  const needHorizontal = dstW !== src.width || x0 !== 0 || x1 !== dstW;
  const needVertical = dstH !== src.height || y0 !== 0 || y1 !== dstH;
  const yboxFirst = vert.bounds[0] ?? 0;
  const yboxLast = (vert.bounds[dstH * 2 - 2] ?? 0) + (vert.bounds[dstH * 2 - 1] ?? 0);
  let img = src;
  if (needHorizontal) {
    for (let i = 0; i < dstH; i += 1) vert.bounds[i * 2] = (vert.bounds[i * 2] ?? 0) - yboxFirst;
    img = horizontalPass(src, yboxFirst, yboxLast - yboxFirst, dstW, horiz);
  }
  if (needVertical) img = verticalPass(img, dstH, vert);
  return img;
}

/**
 * Pillow's Image.crop(box) for an 8-bit image: the box may reach past the source on any side and
 * the outside is filled with black, which the shader pipeline uses where a render's own ground is
 * black (OPENERS.md, "Shader pipeline"; the liquid metal opener's box starts at x -1325).
 */
export function cropPadded(src: GrayImage, box: Box4): GrayImage {
  const [l, t, r, b] = box.map(Math.round) as Box4;
  const width = Math.max(0, r - l);
  const height = Math.max(0, b - t);
  const out = new Uint8Array(width * height);
  const y0 = Math.max(0, t);
  const y1 = Math.min(src.height, b);
  const x0 = Math.max(0, l);
  const x1 = Math.min(src.width, r);
  for (let y = y0; y < y1; y += 1) {
    const srcRow = y * src.width;
    const outRow = (y - t) * width;
    out.set(src.data.subarray(srcRow + x0, srcRow + x1), outRow + (x0 - l));
  }
  return { width, height, data: out };
}

/**
 * The crop box ImageOps.fit computes for a cover fit with the given centering: the largest box
 * of the target aspect inside the source, centered. Fractions are kept, as Pillow keeps them.
 */
export function coverBox(
  srcW: number,
  srcH: number,
  dstW: number,
  dstH: number,
  centering: [number, number] = [0.5, 0.5],
): Box4 {
  const cx = Math.min(1, Math.max(0, centering[0]));
  const cy = Math.min(1, Math.max(0, centering[1]));
  const liveRatio = srcW / srcH;
  const outRatio = dstW / dstH;
  let cropW: number;
  let cropH: number;
  if (liveRatio === outRatio) {
    cropW = srcW;
    cropH = srcH;
  } else if (liveRatio >= outRatio) {
    cropW = outRatio * srcH;
    cropH = srcH;
  } else {
    cropW = srcW;
    cropH = srcW / outRatio;
  }
  const left = (srcW - cropW) * cx;
  const top = (srcH - cropH) * cy;
  return [left, top, left + cropW, top + cropH];
}

/** ImageOps.fit(image, (dstW, dstH), LANCZOS, centering): cover, trimming the long side. */
export function fitCover(
  src: GrayImage,
  dstW: number,
  dstH: number,
  centering: [number, number] = [0.5, 0.5],
): GrayImage {
  return resizeLanczos3(src, dstW, dstH, coverBox(src.width, src.height, dstW, dstH, centering));
}

/** Nearest-neighbour upscale by an integer factor: every cell becomes a k by k square. */
export function scaleNearest(bits: BitImage, k: number): BitImage {
  const f = Math.max(1, Math.round(k));
  const width = bits.width * f;
  const height = bits.height * f;
  const out = new Uint8Array(width * height);
  for (let y = 0; y < height; y += 1) {
    const srcRow = Math.floor(y / f) * bits.width;
    const outRow = y * width;
    for (let x = 0; x < width; x += 1) out[outRow + x] = bits.bits[srcRow + Math.floor(x / f)] ?? 0;
  }
  return { width, height, bits: out };
}

/** Nearest-neighbour upscale of a gray image by an integer factor (the fixture test and thumbs). */
export function scaleNearestGray(gray: GrayImage, k: number): GrayImage {
  const f = Math.max(1, Math.round(k));
  const width = gray.width * f;
  const height = gray.height * f;
  const out = new Uint8Array(width * height);
  for (let y = 0; y < height; y += 1) {
    const srcRow = Math.floor(y / f) * gray.width;
    const outRow = y * width;
    for (let x = 0; x < width; x += 1) out[outRow + x] = gray.data[srcRow + Math.floor(x / f)] ?? 0;
  }
  return { width, height, data: out };
}

File: crates/turboslide-native/src/resample.rs
//! Resampling for the two-tone pipeline (SPEC 5.4, 10): the padded crop, Pillow's Lanczos3 step
//! for step (libImaging/Resample.c, as packages/effects/src/resample.ts reproduces it), the cover
//! fit and the nearest-neighbour upscale.
//!
//! The kernel is `sinc(x) sinc(x / 3)` on |x| < 3, the support scales with the downscale factor,
//! taps are centered at `in0 + (xx + 0.5) * scale` and rounded with `(int)(v + 0.5)`, the
//! coefficients are normalized to one and quantized to 22-bit fixed point, each pass goes through
//! 8-bit storage, and every sum is rounded with a half-unit bias and floored. Floating point is
//! evaluated in the order the TypeScript writes it; `sin` is libm's fdlibm port.

use crate::image::{Bits, Box4, Gray};
use crate::jsmath::js_round;

const SUPPORT: f64 = 3.0;
const PRECISION_BITS: u32 = 32 - 8 - 2;
const PRECISION: f64 = (1u64 << PRECISION_BITS) as f64;
const HALF: i64 = 1 << (PRECISION_BITS - 1);

fn sinc(x: f64) -> f64 {
    if x == 0.0 {
        return 1.0;
    }
    let px = x * std::f64::consts::PI;
    libm::sin(px) / px
}

fn lanczos(x: f64) -> f64 {
    if (-3.0..3.0).contains(&x) {
        sinc(x) * sinc(x / 3.0)
    } else {
        0.0
    }
}

/// The per-axis tap table: `ksize` coefficients per output pixel in `kk`, and for each output
/// pixel its first input index and tap count in `bounds`.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Coeffs {
    pub ksize: usize,
    pub bounds: Vec<i32>,
    pub kk: Vec<i32>,
}

/// Pillow's `precompute_coeffs` plus `normalize_coeffs_8bpc` for one axis.
pub fn precompute_coeffs(in_size: usize, in0: f64, in1: f64, out_size: usize) -> Coeffs {
    let scale = (in1 - in0) / out_size as f64;
    let filterscale = if scale < 1.0 { 1.0 } else { scale };
    let support = SUPPORT * filterscale;
    let ksize = (support.ceil() as usize) * 2 + 1;
    let mut bounds = vec![0i32; out_size * 2];
    let mut kk = vec![0i32; out_size * ksize];
    let mut k = vec![0f64; ksize];
    let ss = 1.0 / filterscale;
    for xx in 0..out_size {
        let center = in0 + (xx as f64 + 0.5) * scale;
        let mut xmin = (center - support + 0.5).trunc() as i64;
        if xmin < 0 {
            xmin = 0;
        }
        let mut xmax = (center + support + 0.5).trunc() as i64;
        if xmax > in_size as i64 {
            xmax = in_size as i64;
        }
        xmax -= xmin;
        let xmax = xmax.max(0) as usize;
        let mut ww = 0.0;
        for x in 0..xmax {
            let w = lanczos(((x as f64 + xmin as f64) - center + 0.5) * ss);
            k[x] = w;
            ww += w;
        }
        for x in 0..xmax {
            if ww != 0.0 {
                k[x] /= ww;
            }
        }
        for x in xmax..ksize {
            k[x] = 0.0;
        }
        for x in 0..ksize {
            let v = k[x];
            kk[xx * ksize + x] = if v < 0.0 {
                (-0.5 + v * PRECISION).trunc() as i32
            } else {
                (0.5 + v * PRECISION).trunc() as i32
            };
        }
        bounds[xx * 2] = xmin as i32;
        bounds[xx * 2 + 1] = xmax as i32;
    }
    Coeffs { ksize, bounds, kk }
}

fn clip8(ss: i64) -> u8 {
    // ss / 2^22 floored: an arithmetic shift floors negatives too, as Math.floor does.
    let v = ss >> PRECISION_BITS;
    if v < 0 {
        0
    } else if v > 255 {
        255
    } else {
        v as u8
    }
}

fn horizontal_pass(src: &Gray, offset: usize, out_h: usize, out_w: usize, c: &Coeffs) -> Gray {
    let mut out = vec![0u8; out_w * out_h];
    for yy in 0..out_h {
        let in_row = (yy + offset) * src.width;
        let out_row = yy * out_w;
        for xx in 0..out_w {
            let xmin = c.bounds[xx * 2] as usize;
            let xmax = c.bounds[xx * 2 + 1] as usize;
            let k_base = xx * c.ksize;
            let mut ss: i64 = HALF;
            for x in 0..xmax {
                ss += src.data[in_row + x + xmin] as i64 * c.kk[k_base + x] as i64;
            }
            out[out_row + xx] = clip8(ss);
        }
    }
    Gray {
        width: out_w,
        height: out_h,
        data: out,
    }
}

fn vertical_pass(src: &Gray, out_h: usize, c: &Coeffs) -> Gray {
    let w = src.width;
    let mut out = vec![0u8; w * out_h];
    for yy in 0..out_h {
        let ymin = c.bounds[yy * 2] as usize;
        let ymax = c.bounds[yy * 2 + 1] as usize;
        let k_base = yy * c.ksize;
        let out_row = yy * w;
        for xx in 0..w {
            let mut ss: i64 = HALF;
            for y in 0..ymax {
                ss += src.data[(y + ymin) * w + xx] as i64 * c.kk[k_base + y] as i64;
            }
            out[out_row + xx] = clip8(ss);
        }
    }
    Gray {
        width: w,
        height: out_h,
        data: out,
    }
}

/// Pillow's `Image.resize(size, LANCZOS, box)` for an 8-bit image. The box is in source pixels
/// and may carry fractions.
pub fn resize_lanczos3(src: &Gray, dst_w: usize, dst_h: usize, bx: Option<Box4>) -> Gray {
    let [x0, y0, x1, y1] = bx.unwrap_or([0.0, 0.0, src.width as f64, src.height as f64]);
    if dst_w == src.width
        && dst_h == src.height
        && x0 == 0.0
        && y0 == 0.0
        && x1 == src.width as f64
        && y1 == src.height as f64
    {
        return src.clone();
    }
    let horiz = precompute_coeffs(src.width, x0, x1, dst_w);
    let mut vert = precompute_coeffs(src.height, y0, y1, dst_h);
    let need_horizontal = dst_w != src.width || x0 != 0.0 || x1 != dst_w as f64;
    let need_vertical = dst_h != src.height || y0 != 0.0 || y1 != dst_h as f64;
    let ybox_first = vert.bounds[0];
    let ybox_last = vert.bounds[dst_h * 2 - 2] + vert.bounds[dst_h * 2 - 1];
    let mut img: Option<Gray> = None;
    if need_horizontal {
        for i in 0..dst_h {
            vert.bounds[i * 2] -= ybox_first;
        }
        img = Some(horizontal_pass(
            src,
            ybox_first as usize,
            (ybox_last - ybox_first) as usize,
            dst_w,
            &horiz,
        ));
    }
    if need_vertical {
        let input = img.as_ref().unwrap_or(src);
        img = Some(vertical_pass(input, dst_h, &vert));
    }
    img.unwrap_or_else(|| src.clone())
}

/// Pillow's `Image.crop(box)`: the box may reach past the source on any side and the outside is
/// filled with black. The edges are rounded with `Math.round`, as the TypeScript does.
pub fn crop_padded(src: &Gray, bx: Box4) -> Gray {
    let l = js_round(bx[0]) as i64;
    let t = js_round(bx[1]) as i64;
    let r = js_round(bx[2]) as i64;
    let b = js_round(bx[3]) as i64;
    let width = (r - l).max(0) as usize;
    let height = (b - t).max(0) as usize;
    let mut out = vec![0u8; width * height];
    let y0 = t.max(0);
    let y1 = b.min(src.height as i64);
    let x0 = l.max(0);
    let x1 = r.min(src.width as i64);
    if x1 > x0 {
        let mut y = y0;
        while y < y1 {
            let src_row = y as usize * src.width;
            let out_row = (y - t) as usize * width;
            let from = src_row + x0 as usize;
            let to = src_row + x1 as usize;
            let at = out_row + (x0 - l) as usize;
            out[at..at + (to - from)].copy_from_slice(&src.data[from..to]);
            y += 1;
        }
    }
    Gray {
        width,
        height,
        data: out,
    }
}

/// The crop box `ImageOps.fit` computes for a cover fit: the largest box of the target aspect
/// inside the source, centered. Fractions are kept.
pub fn cover_box(src_w: usize, src_h: usize, dst_w: usize, dst_h: usize) -> Box4 {
    let (cx, cy) = (0.5, 0.5);
    let live_ratio = src_w as f64 / src_h as f64;
    let out_ratio = dst_w as f64 / dst_h as f64;
    let (crop_w, crop_h) = if live_ratio == out_ratio {
        (src_w as f64, src_h as f64)
    } else if live_ratio >= out_ratio {
        (out_ratio * src_h as f64, src_h as f64)
    } else {
        (src_w as f64, src_w as f64 / out_ratio)
    };
    let left = (src_w as f64 - crop_w) * cx;
    let top = (src_h as f64 - crop_h) * cy;
    [left, top, left + crop_w, top + crop_h]
}

/// `ImageOps.fit(image, (dst_w, dst_h), LANCZOS)`: cover, trimming the long side.
pub fn fit_cover(src: &Gray, dst_w: usize, dst_h: usize) -> Gray {
    resize_lanczos3(
        src,
        dst_w,
        dst_h,
        Some(cover_box(src.width, src.height, dst_w, dst_h)),
    )
}

/// Nearest-neighbour upscale by an integer factor: every cell becomes a k by k square.
pub fn scale_nearest(bits: &Bits, k: f64) -> Bits {
    let f = f64::max(1.0, js_round(k)) as usize;
    let width = bits.width * f;
    let height = bits.height * f;
    let mut out = vec![0u8; width * height];
    for y in 0..height {
        let src_row = (y / f) * bits.width;
        let out_row = y * width;
        for x in 0..width {
            out[out_row + x] = bits.bits[src_row + x / f];
        }
    }
    Bits {
        width,
        height,
        bits: out,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn coefficients_normalize_to_one_unit() {
        let c = precompute_coeffs(100, 0.0, 100.0, 25);
        assert_eq!(c.ksize, 25);
        for xx in 0..25 {
            let sum: i64 = c.kk[xx * c.ksize..(xx + 1) * c.ksize]
                .iter()
                .map(|&v| v as i64)
                .sum();
            // Quantization leaves the sum within a few units of 2^22.
            assert!((sum - (1i64 << 22)).abs() <= 8, "sum {sum} at {xx}");
        }
    }

    #[test]
    fn crop_pads_with_black() {
        let src = Gray {
            width: 2,
            height: 2,
            data: vec![10, 20, 30, 40],
        };
        let out = crop_padded(&src, [-1.0, 0.0, 2.0, 3.0]);
        assert_eq!((out.width, out.height), (3, 3));
        assert_eq!(out.data, vec![0, 10, 20, 0, 30, 40, 0, 0, 0]);
    }

    #[test]
    fn cover_box_trims_the_long_side() {
        let b = cover_box(3200, 1800, 800, 450);
        assert_eq!(b, [0.0, 0.0, 3200.0, 1800.0]);
        let b = cover_box(1000, 1000, 800, 450);
        assert_eq!(b[1], 218.75);
        assert_eq!(b[3], 781.25);
    }

    #[test]
    fn nearest_doubles_every_cell() {
        let bits = Bits {
            width: 2,
            height: 1,
            bits: vec![1, 0],
        };
        let out = scale_nearest(&bits, 2.0);
        assert_eq!(out.bits, vec![1, 1, 0, 0, 1, 1, 0, 0]);
    }
}
